### PR TITLE
Fix mmapbuf::resize() and re-enable aligned resize test

### PR DESCRIPTION
**Pull request by tobim**
The algorithm to expand a mmaped region is buggy and has already triggered segfaults on linux systems.

### CODE REVIEW

File: libvast/src/detail/mmapbuf.cpp [hunk 1]
```diff
@@ -105,6 +105,15 @@ bool mmapbuf::resize(size_t new_size) {
   // Resize the underlying file, if available.
   if (fd_ != -1 && ftruncate(fd_, new_size) < 0)
     return false;
+#if __linux__
```
Comment (Neverlord) [inline libvast/src/detail/mmapbuf.cpp]: We wrap the various platform-dependent defines: `#ifdef VAST_LINUX` (make sure `config.hpp` is included).

File: libvast/src/detail/mmapbuf.cpp [hunk 2]
```diff
@@ -105,6 +105,15 @@ bool mmapbuf::resize(size_t new_size) {
...
+  int flags = MREMAP_MAYMOVE;
+  auto* map = mremap(map_, size_, new_size, flags);
```
Comment (Neverlord) [inline libvast/src/detail/mmapbuf.cpp]: `auto map` (the `*` is redundant).

**Comment by Neverlord** [comment]
@tobim please wait before pushing the changes until https://github.com/vast-io/vast/pull/229 is merged (and then rebase to `master` before pushing to get a clean build).

File: libvast/src/detail/mmapbuf.cpp [hunk 3]
```diff
@@ -95,7 +95,7 @@ size_t mmapbuf::size() const {
 bool mmapbuf::resize(size_t new_size) {
   // Also fail if we created a private mapping of a file not opened RW.
   // For now, users cannot control these aspects.
-  if (!map_ || offset_ > new_size)
+  if (!map_)
```
Comment (mavam) [inline libvast/src/detail/mmapbuf.cpp]: This appears to be an orthogonal fix. Technically, it should go into a separate commit.

File: libvast/src/detail/mmapbuf.cpp [hunk 4]
```diff
@@ -131,7 +140,7 @@ bool mmapbuf::resize(size_t new_size) {
     // If the current mapping is a multiple of the page size, we can try (1).
     if (size_ % page_size() == 0) {
       auto flags = flags_ | MAP_FIXED;
-      auto map = mmap(map_ + size_, new_size - size_, prot_, flags, fd_, 0);
+      auto map = mmap(map_ + size_, new_size - size_, prot_, flags, fd_, offset_ + size_);
```
Comment (mavam) [inline libvast/src/detail/mmapbuf.cpp]: This change seems likewise orthogonal to the Linux fix.

Also seems like we are exceeding 80 characters here; please wrap accordingly.

Comment (Neverlord) [reply libvast/src/detail/mmapbuf.cpp]: > exceeding 80 characters

Checking basic coding style rules (80 characters, no unnecessary vertical whitespaces, 2-space indentation, ...) sounds like something we should automate.

File: libvast/src/detail/mmapbuf.cpp (hunk 2 above)
Comment (tobim) [reply libvast/src/detail/mmapbuf.cpp]: In this case you are right, but there is a subtle nuisance of c++s auto type deduction rules that we have to keep in mind: if you want to assign the the returned pointer to a `const` variable, you have to use `const auto*`. `const auto` would result in `T* const` instead. For this reason, I prefer using `auto*` and `const auto*` instead of `auto` and `const auto*`.

I will change the line here nonetheless to comply with the coding style, but I wanted to mention the issue.

File: libvast/src/detail/mmapbuf.cpp (hunk 4 above)
Comment (mavam) [reply libvast/src/detail/mmapbuf.cpp]: Absolutely. We're not at a point yet where we can run `clang-format` over every PR, but we should consider basic linting.

**Comment by tobim** [comment]
@Neverlord branch is updated with your requested changes and an additional fix for platforms other than linux. Rebase is done as well.

File: libvast/test/mmapbuf.cpp [hunk 5]
```diff
@@ -61,25 +61,39 @@ TEST(memory-mapped streambuffer) {
...
+TEST(memory-mapped streambuffer aligned resize) {
+  auto filename = directory / "aligned";
+  aligned_resize_test_impl(filename, detail::page_size());
+}
+
+TEST(memory-mapped streambuffer aligned resize large) {
+  auto filename = directory / "aligned_large";
+  auto page_size = detail::page_size();
+  auto hundred_mb = (100 * (1 << 20));
+  auto size = hundred_mb - (hundred_mb % page_size);
```
Comment (Neverlord) [inline libvast/test/mmapbuf.cpp]: This should raise a warning for mixing signed and unsigned integers. `hundred_mb` is an `int`, but `page_size` is a `size_t`. Since `aligned_resize_test_impl` expects a `size_t` as second argument, I suggest to convert `hundred_mb` to a `size_t`.

File: libvast/test/mmapbuf.cpp [hunk 6]
```diff
@@ -61,25 +61,39 @@ TEST(memory-mapped streambuffer) {
...
   REQUIRE(sb.data() != nullptr);
-  CHECK_EQUAL(sb.size(), page_size);
+  CHECK_EQUAL(sb.size(), size);
+  sb.sputn("Here be content", 15);
   // Aligned resizing.
-  REQUIRE(sb.resize(page_size * 2));
-  CHECK_EQUAL(sb.size(), page_size * 2);
+  REQUIRE(sb.resize(size * 2));
+  CHECK_EQUAL(sb.size(), size * 2);
+  CHECK_EQUAL(std::string(sb.data()+3, 9), "e be cont");
```
Comment (mavam) [inline libvast/test/mmapbuf.cpp]: Style: whitespace around binary operators.

File: libvast/test/mmapbuf.cpp [hunk 7]
```diff
@@ -61,25 +61,39 @@ TEST(memory-mapped streambuffer) {
   CHECK_EQUAL(cur, sb.size()); // we're at the end!
 }
 
-TEST(memory-mapped streambuffer aligned resize) {
-  auto filename = directory / "aligned";
-  auto page_size = detail::page_size();
-  detail::mmapbuf sb{filename.str(), page_size};
+void aligned_resize_test_impl(const vast::path& filename, size_t size) {
```
Comment (mavam) [inline libvast/test/mmapbuf.cpp]: We usually wrap functions in the implementation into an anonymous namespace:
```cpp
namespace {

void f() {
  // ...
}

} // namespace <anonymous>

**Comment by mavam** [comment]
> Do you mind if I add a similar flag to VAST's configure?

Not at all, please go ahead.

> Mid-term we should strive for zero-warnings builds and enforce this via Jenkins.

Absolutely!
